plutus-use-cases/src/Language/PlutusTx/Coordination/Contracts/Future.hs

```diff
@@ -73,230 +151,453 @@ their margin. If either party fails to make a margin payment then the contract
+
+settleEarly
+    :: ( HasEndpoint "settle-early" (OracleValue Value) s
+       , HasBlockchainActions s
+       , AsFutureError e
+       )
+    => Future
+    -> FutureOwners
+    -> Contract s e ()
+settleEarly future@Future{ftDeliveryDate} ftos = do
+    ov <- endpoint @"settle-early"
+    let address = futureAddress future ftos
+        spotPrice = ovValue ov
+        minMargin = requiredMargin future spotPrice
+    unspentOutputs <- utxoAt address
+    MarginAccounts{ftsShortMargin, ftsLongMargin} <- currentBalances future ftos
+    
+    violatingRole <- 
+        if ftsShortMargin `lt` minMargin then pure Short
+        else if ftsLongMargin `lt` minMargin then pure Long
+        else throwing_ _MarginRequirementsNotViolated
+    
+    let payment = case violatingRole of
+                    Long -> Payouts{payoutsLong=mempty, payoutsShort = ftsShortMargin + ftsLongMargin }
+                    Short -> Payouts{payoutsLong=ftsShortMargin+ftsLongMargin, payoutsShort = mempty }
+        tx = payoutsTx payment ftos
+             Haskell.<> Typed.collectFromScript unspentOutputs (scriptInstance future ftos) (SettleEarly ov)
+    void $ writeTx tx
+
+-- | Determine the current 'MarginAccounts' by looking at the unspent output
+--   of the contract.
+currentBalances
+    :: ( HasUtxoAt s
+       , AsFutureError e)
+    => Future
+    -> FutureOwners
+    -> Contract s e MarginAccounts
+currentBalances future ftos = do
+    let address = futureAddress future ftos
+    unspentOutputs <- utxoAt address
+    theOutput <- case toList (AM.outRefMap unspentOutputs) of
+                    []  -> throwing_ _FutureUtxoNotFound
+                    [x] -> pure x
+                    _   -> throwing_ _FutureUtxoNotUnique
+    Ledger.DataScript theData <- maybe (throwing_ _FutureUtxoNoData) pure (Ledger.txOutTxData theOutput)
+    theState <- maybe (throwing _FutureUtxoFromDataFailed theData) pure (PlutusTx.fromData theData)
+    case theState of
+        Running accounts -> pure accounts
+        Finished         -> throwing_ _FutureTerminated
+
+increaseMargin
+    :: ( HasEndpoint "increase-margin" (Value, Role) s
+       , HasUtxoAt s
+       , HasWriteTx s
+       , AsFutureError e
+       )
+    => Future
+    -> FutureOwners
+    -> Contract s e ()
+increaseMargin future ftos = do
+    (value, role) <- endpoint @"increase-margin"
+    let address = futureAddress future ftos
+    unspentOutputs <- utxoAt address
+    currentState <- currentBalances future ftos
+    let newMarginAccounts = adjustMargin role value currentState
+        dataScript = Ledger.DataScript (PlutusTx.toData $ Running newMarginAccounts)
+        tx =
+            Tx.collectFromScript unspentOutputs (validatorScript future ftos) (Ledger.RedeemerScript $ PlutusTx.toData $ AdjustMargin role value)
+            Haskell.<> payToScript (totalMargin newMarginAccounts) address dataScript
+    void (withContractError (review _MarginAdjustmentFailed) (writeTxSuccess tx))
+
+join
```

Comment:
Can we not call this `join`? I automatically assume that's `Control.Monad.join`.